### PR TITLE
ICU-20775 Complete set of toDecimalNumber functions in ICU4C

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20775
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by markusicu** [comment]
Please adjust according to the API discussion.

**Comment by sffc** [comment]
The first force-push was the update according to API review.  The second force-push was rebasing against master in order to resolve a merge conflict involving the `U_DRAFT` to `U_CAPI` change.

File: icu4c/source/test/intltest/numbertest_range.cpp [hunk 1]
```diff
@@ -860,6 +861,37 @@ void NumberRangeFormatterTest::toObject() {
...
+        .numberFormatterBoth(NumberFormatter::with().unit(USD));
+
+    // Range of numbers
+    {
+        FormattedNumberRange range = lnf.formatFormattableRange(1, 5, status);
+        assertEquals("Range: Formatted string should be as expected",
+            u"$1.00 \u2013 $5.00",
+            range.toString(status));
+        auto decimalNumbers = range.getDecimalNumbers<std::string>(status);
+        // NOTE: DecNum doesn't retain trailing zeros. Is that a problem?
```
Comment (markusicu) [inline icu4c/source/test/intltest/numbertest_range.cpp:877]: I don't know the DecNum spec, but this means that the FormattedNumberRange getter doesn't quite return a string with the fidelity that the getter docs promise.

File: icu4c/source/i18n/number_capi.cpp [hunk 2]
```diff
@@ -196,6 +197,23 @@ unumf_resultGetAllFieldPositions(const UFormattedNumber* uresult, UFieldPosition
     result->fData.getAllFieldPositions(fpih, *ec);
 }
 
+U_DRAFT int32_t U_EXPORT2
```
Comment (markusicu) [inline icu4c/source/i18n/number_capi.cpp:200]: U_CAPI

**Review by markusicu — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/numrange_capi.cpp [hunk 3]
```diff
@@ -138,6 +139,40 @@ unumrf_resultGetIdentityResult(
...
+        char* dest,
+        int32_t destCapacity,
+        UErrorCode* ec) {
+    auto* result = UFormattedNumberRangeApiHelper::validate(uresult, *ec);
+    if (U_FAILURE(*ec)) {
+        return 0;
+    }
+    DecNum decnum;
+    return result->fData.quantity1
+        .toDecNum(decnum, *ec)
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/numrange_capi.cpp:154]: I re-formatted it using clang-format.  It decided to put `.toDecNum` on the first line and keep the remaining lines with leading dots.

File: icu4c/source/test/intltest/numbertest_range.cpp (hunk 1 above)
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_range.cpp:877]: I opened a ticket and added logKnownIssue to investigate further.

File: icu4c/source/i18n/numrange_capi.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4c/source/i18n/numrange_capi.cpp]: That doesn't make it any less strange :-)
But, "optional".

**Review by markusicu — COMMENTED** [review]
lgtm tnx -- please squash

**Review by markusicu — APPROVED** [review]
lgtm

**Pull request merged**